modules/porous_flow/src/fvkernels/FVPorousFlowHeatAdvection.C

```diff
@@ -0,0 +1,103 @@
+  InputParameters params = FVFluxKernel::validParams();
+  RealVectorValue g(0, 0, -9.81);
+  params.addParam<RealVectorValue>("gravity", g, "Gravity vector. Defaults to (0, 0, -9.81)");
+  params.addRequiredParam<UserObjectName>("PorousFlowDictator",
+                                          "The PorousFlowDictator UserObject");
+  params.set<unsigned short>("ghost_layers") = 2;
+  params.addClassDescription("Heat flux advected by the fluid");
+  return params;
+}
+
+FVPorousFlowHeatAdvection::FVPorousFlowHeatAdvection(const InputParameters & params)
+  : FVFluxKernel(params),
+    _dictator(getUserObject<PorousFlowDictator>("PorousFlowDictator")),
+    _num_phases(_dictator.numPhases()),
+    _density(getADMaterialProperty<std::vector<Real>>("PorousFlow_fluid_phase_density_qp")),
+    _density_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_fluid_phase_density_qp")),
+    _viscosity(getADMaterialProperty<std::vector<Real>>("PorousFlow_viscosity_qp")),
+    _viscosity_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_viscosity_qp")),
+    _enthalpy(getADMaterialProperty<std::vector<Real>>("PorousFlow_fluid_phase_enthalpy_qp")),
+    _enthalpy_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_fluid_phase_enthalpy_qp")),
+    _relperm(getADMaterialProperty<std::vector<Real>>("PorousFlow_relative_permeability_qp")),
+    _relperm_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_relative_permeability_qp")),
+    _permeability(getADMaterialProperty<RealTensorValue>("PorousFlow_permeability_qp")),
+    _permeability_neighbor(
+        getNeighborADMaterialProperty<RealTensorValue>("PorousFlow_permeability_qp")),
+    _pressure(getADMaterialProperty<std::vector<Real>>("PorousFlow_porepressure_qp")),
+    _pressure_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_porepressure_qp")),
+    _grad_p(getADMaterialProperty<std::vector<RealGradient>>("PorousFlow_grad_porepressure_qp")),
+    _gravity(getParam<RealVectorValue>("gravity"))
+{
+}
+
+ADReal
+FVPorousFlowHeatAdvection::computeQpResidual()
+{
+  ADReal flux = 0.0;
+  ADRealGradient pressure_grad;
+  ADRealTensorValue mobility;
+
+  for (unsigned int p = 0; p < _num_phases; ++p)
+  {
+    // If we are on a boundary face, use the reconstructed gradient computed in _grad_p
+    if (onBoundary(*_face_info))
+    {
+      const auto & gradp = -_grad_p[_qp][p];
+      pressure_grad = gradp + _density[_qp][p] * _gravity;
+
+      mobility = _enthalpy[_qp][p] * _relperm[_qp][p] * _permeability[_qp] * _density[_qp][p] /
+                 _viscosity[_qp][p];
+    }
+    else
+    {
+      // If we are on an internal face, calculate the gradient explicitly
+      const auto & p_elem = _pressure[_qp][p];
+      const auto & p_neighbor = _pressure_neighbor[_qp][p];
+
+      const auto gradp = (p_elem - p_neighbor) * _face_info->eCN() / _face_info->dCNMag();
```

Comment:
that s one way to do it. This is possibly a cheap way. We have machinery in moose to go GreenGauss on both elem in neighbor then interpolate the gradient of the pressure if you want to try

Reply:
INSFV has orthogonality and skewness corrections implemented too if that is of interest

Reply:
I'll look into these and see if it makes a difference. Thanks for letting me know about these!

Reply:
sounds good. Please re-ping me for review once you have decided what you want to use

Reply:
you already have `_grad_p[_qp]` available in this fvkernel just up there. From looking at the other thread it would just be a simple substitution with this to try orthogonality correction

Reply:
So do you mean that I don't have to explicitly compute gradp like above, and using _grad_p[_qp] will automatically provide the correction?

Reply:
ah no I got confused. 
You already have grad_p which is an elemental gradient. 

If you want grad_p on the face you will need to call the adgradSln routine with the faceinfo argument

Reply:
Hey @GiudGiud - I'm going to have to think about these additional features like skewness correction etc and how to modify the other FV kernels in porous flow. In the meantime, I have someone wanting to use this who doesn't need those features. Can I get this merged in and I'll discuss further with you guys how to implement those features?

Reply:
I should say - I think I have addressed everything else in your original review. Thanks for the comments and ideas!

Reply:
I m giving a MOOSE training for the next two days. Can it wait until Thursday? If not we can re-assign the reviewer it's nbd

Reply:
Thursday is fine!